### PR TITLE
Scan: support cropping.

### PR DESCRIPTION
**Pull request by JoepVanlier**
**Why this PR?**
We offer the functionality to crop `CorrelatedStacks`, this PR introduces the same behaviour for `Scans`.

One thing to note is that I had to overhaul the slicing by frame to achieve the correct behaviour. Reason being that the old slicing relied on always reconstructing the array. Slicing by frame resetting the crop seems unacceptable to me, as I could see that to be a regular use case (crop first, select frames later).

This comes at a penalty, namely that reconstruction takes place for the entire image array when selecting a range of frames, whereas before, if you sliced frames, you would only reconstruct what you use. In practice, I think for most use cases, the change is still warranted for two reasons:

1. I think it is rare that you know exactly which frames you want in advance, meaning that you will end up reconstructing the whole array at first anyway. In this use case, the current implementation is actually _faster_ since it'll reuse the cache from that first reconstruction for your subsequent crops.
2. The slicing time is dominated by obtained the `frame_timestamp_ranges()` anyhow (which involves two `timestamp` reconstructions).

Note, `crop_by_pixels` is specifically made public API because it is used by some of our cropping widgets.

### CODE REVIEW

File: lumicks/pylake/tests/test_imaging_confocal/test_scan.py [hunk 1]
```diff
@@ -328,11 +328,11 @@ def test_single_frame_times(dim_x, dim_y, line_padding, start, dt, samples_per_p
     [
         (5, 6, 3, 3, 14, 4, 4),
         (3, 4, 4, 60, 1592916040906356300, 12800, 30),
-        (3, 2, 2, 60, 1592916040906356300, 12800, 3000),
+        (3, 2, 3, 60, 1592916040906356300, 12800, 3000),
     ],
 )
 def test_multiple_frame_times(dim_x, dim_y, frames, line_padding, start, dt, samples_per_pixel):
-    img = np.ones((dim_x, dim_y, frames))
+    img = np.ones((frames, dim_x, dim_y))
```
Comment (rpauszek) [inline lumicks/pylake/tests/test_imaging_confocal/test_scan.py:337]: hmm somewhat concerning that this passed even with an incorrect shape definition before

File: lumicks/pylake/tests/test_imaging_confocal/test_scan.py [hunk 2]
```diff
@@ -478,3 +479,175 @@ def test_plot_single_channel_percentile_color_adjustment(test_scans):
...
+
+    assert_equal(multi_frame[1:3][:, 1:3, 2:3], multi_frame[1:3, 1:3, 2:3])
+    assert_equal(multi_frame[1][:, 1:3, 2:3], multi_frame[1, 1:3, 2:3])
+    assert_equal(multi_frame[:, 1:3, 2:3][3], multi_frame[3, 1:3, 2:3])
+    assert_equal(multi_frame[:, 1:, 2:][3], multi_frame[3, 1:, 2:])
+    assert_equal(multi_frame[:, :3, :3][3], multi_frame[3, :3, :3])
+    assert_equal(single_frame[0][:, 1:5, 2:6][:, 1:3, 2:8], single_frame[0, 2:4, 4:6])
+
+
+def test_error_handling_multidim_indexing():
```
Comment (rpauszek) [inline lumicks/pylake/tests/test_imaging_confocal/test_scan.py:609]: missing a test for slicing with `list`, which should raise `IndexError("Only slicing by int or slice is allowed to select frames")`, but connected with my comment about this above

**Review by rpauszek — CHANGES_REQUESTED** [review]
Really nice work especially considering there's so many different combinations/corner cases to test. There's some small comments about code clarity but honestly I'm not sure if my suggestions are any better; I think this block is a necessarily complicated piece of code.

However, I did run across a case where you don't get the expected exception (also missing a test) that should be corrected

File: lumicks/pylake/scan.py [hunk 3]
```diff
@@ -31,32 +32,134 @@ def __init__(self, name, file, start, stop, metadata):
...
-            if item.step is not None:
-                raise IndexError("Slice steps are not supported")
+        def check_item(item, allow_int):
+            """We don't allow slice steps, and we don't allow slicing with lists. Slicing with
+            single integer value is only allowed for frames."""
+            if isinstance(item, slice):
+                if item.step is not None:
+                    raise IndexError("Slice steps are not supported when indexing")
+                return item
+            else:
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/scan.py:145]: nice one! 👍

File: lumicks/pylake/scan.py [hunk 4]
```diff
@@ -31,32 +32,134 @@ def __init__(self, name, file, start, stop, metadata):
...
-                raise NotImplementedError("Slice is empty.")
+                if isinstance(item, int):
+                    return item
 
-            start, stop = sliced_ts_ranges[0][0], sliced_ts_ranges[-1][-1]
-            num_frames = len(sliced_ts_ranges)
-        else:
-            start, stop = ts_ranges[item]
-            num_frames = 1
+                raise IndexError("Only slicing by int or slice is allowed to select frames")
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/scan.py:152]: Right, that's a good point. I restructured it a bit and produce the correct error message now.

```python
            if isinstance(item, slice):
                if item.step is not None:
                    raise IndexError("Slice steps are not supported when indexing")
                return item

            if isinstance(item, int):
                if not allow_int:
                    raise IndexError("Scalar indexing is not supported for spatial coordinates")
                else:
                    return item

            raise IndexError(f"Slicing by {type(item).__name__} is not allowed.")
```

I think code-wise it's also a bit less confusing, since now you immediately see which type is attempted at each block.

File: lumicks/pylake/scan.py [hunk 5]
```diff
@@ -31,32 +32,134 @@ def __init__(self, name, file, start, stop, metadata):
...
+            raise NotImplementedError("Slice is empty.")
+
+        return result
+
     def __repr__(self):
         name = self.__class__.__name__
         return f"{name}(pixels=({self.pixels_per_line}, {self.lines_per_frame}))"
 
     def __getitem__(self, item):
         """All indexing is in frames"""
```
[the remark this answers is not in the document]
Comment (rpauszek) [reply lumicks/pylake/scan.py]: also noticed this needs to be done for `CorrelatedStack`; maybe worthwhile just piggybacking it here?

File: lumicks/pylake/tests/test_imaging_confocal/test_scan.py (hunk 2 above)
Comment (JoepVanlier) [reply lumicks/pylake/tests/test_imaging_confocal/test_scan.py:609]: Indeed. Added a slightly more general test which just flags that whatever type you sliced with isn't allowed.

File: lumicks/pylake/scan.py [hunk 6]
```diff
@@ -31,32 +32,134 @@ def __init__(self, name, file, start, stop, metadata):
...
+        """
+        slices = tuple((frame_axis, *spatial_axes)) if self.num_frames > 1 else tuple(spatial_axes)
+
+        if isinstance(frame_axis, int):
+            frame_axis = frame_axis if frame_axis >= 0 else self.num_frames + frame_axis
+            if not 0 <= frame_axis < self.num_frames:
+                raise IndexError("Frame index out of range")
+        else:
+            if not frame_axis.indices(self.num_frames):
+                raise IndexError("Frame indices need to be within scan range")
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/scan.py]: Correct. I think this was a mishap when I was refactoring this code. This was fixed again in the latest commit where I check the range of it. That one behaves correctly and is falsy when it's empty.

File: lumicks/pylake/scan.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/scan.py]: Sure. I'll piggyback it.

File: lumicks/pylake/scan.py [hunk 7]
```diff
@@ -31,32 +32,134 @@ def __init__(self, name, file, start, stop, metadata):
...
+        return self._scan_with_sliced_factories(frame_axis=slice(None), spatial_axes=spatial_axes)
+
+    def _scan_with_sliced_factories(self, frame_axis, spatial_axes) -> "Scan":
+        """Return a scan with sliced factories
+
+        Parameters
+        ----------
+        frame_axis : slice or int
+            How to slice the frame axis
+        spatial_axes : tuple of slice
```
[the remark this answers is not in the document]
Comment (rpauszek) [reply lumicks/pylake/scan.py:59]: `spatial_axes : () or (slice) or (slice, slice)` ?

Might seem like a bit much, but at least it's explicit. If possible the type annotation should be valid code, no?

File: lumicks/pylake/scan.py [hunk 8]
```diff
@@ -31,32 +32,166 @@ def __init__(self, name, file, start, stop, metadata):
...
+            scan[5]  # Gets the 6th frame of the scan (0 is the first).
+            scan[1:5]  # Get scan frames 1, 2, 3 and 4.
+            scan[:, 10:50, 20:50]  # Gets all frames, but crops them from 10 to 50 and 20 to 50.
+            scan[:, 10:50]  # Gets all frames, but crops one spatial dimension from 10 to 50.
+            scan[5, 10:20, 10:20]  # Obtains the 6th frame and crops it.
+
+            scan[[1, 3, 4]]  # Produces an error, lists are not allowed.
+            scan[1:5:2]  # Produces an error, steps are not allowed.
+            scan[1, 3, 5]   # Produces an error, indices are not allowed for the spatial dimensions.
+            scan[1, 3:5:2]  # Produces an error, steps are not allowed when slicing.
```
Comment (rpauszek) [inline lumicks/pylake/scan.py:161]:
```suggestion
            scan[5]  # Gets the 6th frame of the scan (0 is the first).
            scan[1:5]  # Get scan frames 1, 2, 3 and 4.
            scan[:, 10:50, 20:50]  # Gets all frames cropped from row 11 to 50 and column 21 to 50.
            scan[:, 10:50]  # Gets all frames and all columns, but crops from row 11 to 50.
            scan[5, 10:20, 10:20]  # Obtains the 6th frame and crops it.

            scan[[1, 3, 4]]  # Produces an error, lists are not allowed.
            scan[1:5:2]  # Produces an error, steps are not allowed.
            scan[1, 3, 5]   # Produces an error, integer indices are not allowed for the spatial dimensions.
            scan[1, 3:5:2]  # Produces an error, steps are not allowed when slicing.
```

File: lumicks/pylake/scan.py [hunk 9]
```diff
@@ -31,32 +32,166 @@ def __init__(self, name, file, start, stop, metadata):
...
+            single integer value is only allowed for frames."""
+            if isinstance(item, slice):
+                if item.step is not None:
+                    raise IndexError("Slice steps are not supported when indexing")
+                return item
+
+            if isinstance(item, int):
+                if not allow_int:
+                    raise IndexError("Scalar indexing is not supported for spatial coordinates")
+                else:
```
Comment (rpauszek) [inline lumicks/pylake/scan.py:175]: could actually get rid of this too

File: lumicks/pylake/tests/test_imaging_confocal/test_scan.py [hunk 10]
```diff
@@ -478,3 +481,189 @@ def test_plot_single_channel_percentile_color_adjustment(test_scans):
...
+
+def test_slice_by_list_disallowed(test_scans):
+    with pytest.raises(IndexError, match="Slicing by list is not allowed"):
+        test_scans["fast Y slow X multiframe"][[0, 1], :, :]
+
+    with pytest.raises(IndexError, match="Slicing by list is not allowed"):
+        test_scans["fast Y slow X multiframe"][:, [0, 1], :]
+
+    class Dummy:
+        pass
```
Comment (rpauszek) [inline lumicks/pylake/tests/test_imaging_confocal/test_scan.py:666]: 👌

File: lumicks/pylake/scan.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rpauszek) [reply lumicks/pylake/scan.py]: be careful what you wish for 😆

**Review by rpauszek — APPROVED** [review]
looks great! left a few minor suggestions; I'll leave it up to you which if any you'd like to implement

File: lumicks/pylake/scan.py [hunk 11]
```diff
@@ -31,32 +32,166 @@ def __init__(self, name, file, start, stop, metadata):
...
-        ts_ranges = self.frame_timestamp_ranges(include_dead_time=True)
+        """Obtain scan frames or cropped scans using indexing.
 
-        if isinstance(item, slice):
-            if item.step is not None:
-                raise IndexError("Slice steps are not supported")
+        The first index refers to the scan frames and should always be in frames. For the frame
+        number, both indexing and slicing are allowed, but using steps or slicing by list is not.
+        For the spatial dimensions only slicing by space (without steps) is allowed, and should be
+        performed in pixel values.
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/scan.py:141]: 🙌

File: lumicks/pylake/scan.py (hunk 9 above)
Comment (JoepVanlier) [reply lumicks/pylake/scan.py:175]: 👍

File: lumicks/pylake/tests/test_imaging_confocal/test_scan.py (hunk 1 above)
Comment (JoepVanlier) [reply lumicks/pylake/tests/test_imaging_confocal/test_scan.py:337]: Indeed. Was a flawed test. Never tested against the actually requested frame count. Fixed now though. It'll fire if you change it back to the error.
